### PR TITLE
Layer specific control bars

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description

This PR implements a refactor described in #107 where now individual layers determine what is displayed in the "control bars" panel - this is done using stacked widgets. The image layer displays a range slider for controlling clims, the markers layer displays a button group for determining interactivity.

The layers are now also allowed to determine their own cursors (from an acceptable dictionary) and statusbar updates. There is parts of the code that have not got clean MVC separation but we can work on that either in this PR or a followup PR. I've also disabled the keyboard shortcuts, but can reimplement them either in this PR of a followup. Now individual layers will be able to define their own keyboard shortcuts (so different ones could be used for an image layer or a markers layer). I also want to improve how mouse events (and then keyboard events) are forwarded on to the layers from the viewer.

Something else that is becoming clear is that the splitting of viewer into controls and dimensions is maybe causing more headaches than it is solving - for example now controls is very minimal and mainly modifies things on the viewer, so certainly that logic can be merged back into viewer. I will do that in a followup PR though.

Here is a gif of the new control bars in action:
layer_control_bars2

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# References

# How has this been tested?

- [x] test using `examples/layers.ipynb`

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: gui/layers/_markers_layer/model.py [hunk 1]
```diff
@@ -236,6 +241,38 @@ def scaling(self, scaling: bool) -> None:
...
+    @property
+    def mode(self):
+        """None, str: Interactive mode
+        """
+        return self._mode
+
+    @mode.setter
+    def mode(self, mode):
+        # if mode == self.mode:
+        #     return
```
Comment (kne42) [inline gui/layers/_markers_layer/model.py]: Why the commented code?

File: gui/layers/_markers_layer/model.py [hunk 2]
```diff
@@ -236,6 +241,38 @@ def scaling(self, scaling: bool) -> None:
...
+            self._mode = mode
+        elif mode == 'pan/zoom':
+            self.cursor = 'standard'
+            self.interactive = True
+            self.help = ''
+            self.status = mode
+            self._mode = mode
+        else:
+            raise ValueError("Mode not recongnized")
+        #self.refresh()
```
Comment (kne42) [inline gui/layers/_markers_layer/model.py]: ^See above

File: gui/layers/_markers_layer/model.py [hunk 3]
```diff
@@ -262,6 +299,16 @@ def _refresh(self):
         self._need_display_update = True
         self._update()
 
+    # def _set_mode(self, mode):
```
Comment (kne42) [inline gui/layers/_markers_layer/model.py]: ^

File: gui/layers/_markers_layer/model.py [hunk 4]
```diff
@@ -391,55 +438,66 @@ def get_value(self, position, indices):
...
-        else:
-
+        if index is not None:
             self._size = delete(self._size, index, axis=0)
             self.data = delete(self.data, index, axis=0)
             self._selected_markers = None
 
-    def move(self, position, indices):
-        """Moves object at given mouse position and set of indices.
+    def _move(self, coord):
```
Comment (kne42) [inline gui/layers/_markers_layer/model.py:460]: What's your reasoning for making this private?

File: gui/layers/_markers_layer/view/controls.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        layout.addWidget(self.addition_button)
+        layout.addWidget(self.panzoom_button)
+        layout.addStretch(0)
+        self.setLayout(layout)
+        self.setMouseTracking(True)
+
+    def mouseMoveEvent(self, event):
+        self.layer.viewer.status = self.layer.status
+
+    # def on_key_press(self, event):
```
Comment (kne42) [inline gui/layers/_markers_layer/view/controls.py]: What was all of this for?

**Review by kne42 — APPROVED** [review]
😍 Looks fantastic thank you! Left a few comments asking about commented code, privatization of some methods, etc.

**Comment by kne42** [comment]
One added comment: why is locking pan/zoom specific to the markers layer? Wouldn't this be useful as a general feature? Also, could we think up a more explicit name than `Controls`?

File: gui/layers/_markers_layer/model.py (hunk 1 above)
Comment (sofroniewn) [reply gui/layers/_markers_layer/model.py]: will fix, my mistake

File: gui/layers/_markers_layer/model.py (hunk 2 above)
Comment (sofroniewn) [reply gui/layers/_markers_layer/model.py]: did not need, so now removed

File: gui/layers/_markers_layer/model.py (hunk 3 above)
Comment (sofroniewn) [reply gui/layers/_markers_layer/model.py]: deleted

File: gui/layers/_markers_layer/model.py (hunk 4 above)
Comment (sofroniewn) [reply gui/layers/_markers_layer/model.py:460]: This now only gets called from within the markers layer. It isn't quite generic enough at this point that I'd call if from a notebook, but maybe that will change when we do the shapes stuff and improve this sort of editing / moving api. Maybe we revisit then

File: gui/layers/_markers_layer/view/controls.py (hunk 5 above)
Comment (sofroniewn) [reply gui/layers/_markers_layer/view/controls.py]: This was how we did all the key press stuff before, like entering annotation mode or space bar to get back into zoom mode. I will reimplement, can either do in this PR or the next one.

**Comment by sofroniewn** [comment]
Locking pan/zoom can be done with any layer with the `interactive` property, or done with the viewer `interactive` property so can be used by other layer types. Markers is the only one that does this right now / has an explicit button in the gui for it. We can think about how we'd expose this option in a consistent way for other layer types too.

I think `Controls` is going to disappear and get absorbed into `Viewer` so we might not even need a new name. I'll do this in the next PR though

**Comment by sofroniewn** [comment]
I have made a couple more improvements to this PR, like better forwarding of mouse / keypress events onto the individual layers. Each layer type now has complete flexibility on how it handles mouse and keypress events.

I've also absorbed the `Controls` component into the `Viewer` and added a separate view to the viewer for it. This reduces some of the modularity overhead we had taken on.

Now is a good time to review the PR. There's still some cleaning up / separation of concerns stuff to do here (improving our adopting of MVC like paradigms) but again I think we are stepping in the right direction.

**Comment by freeman-lab** [comment]
I've tested this out and reviewed it from a functionality and UX perspective (didn't look in too much detail at the code, but I think @kne42 covered that!) -- overall looking really good!

The idea of control elements (sliders, buttons, etc.) being enabled / disabled dependent on the layer type selected seems great to have implemented.

From a layout perspective, I'm still a bit unsure whether all elements should be at the same location and literally swap in and out (as in this PR), vs having at least a portion of the UI and a small subset of buttons always present  -- consistency can help the layout feel "grounded". But, in the current set of layer types, I agree that clims make sense only for the image type, and the buttons only make sense for the marker type, so the current approach feels right. I'd be open to exploring this again, especially if we can get some user feedback.

But, I'd say, merge the PR as is! Can easily iterate layout later on, and again, the basic concept of linking the controls to the layer type seems like the direction to move in.

**Review by freeman-lab — APPROVED** [review]

**Comment by royerloic** [comment]
Love it!

**Comment by jni** [comment]
Can I request that we have `example/demo.py` always working *and* make `.py` files the primary testbed, rather than Jupyter notebooks? One of the main reasons for Napari's existence is that I hate having to launch a browser to run Python programs. =)

Anyway, when I run `python example/demo.py` I get the following error, which can't be good:

```pytb
 $ python examples/demo.py
/Users/jni/conda/envs/36/lib/python3.6/site-packages/vispy/visuals/isocurve.py:22: UserWarning: VisPy is not yet compatible with matplotlib 2.2+
  warnings.warn("VisPy is not yet compatible with matplotlib 2.2+")
WARNING: Traceback (most recent call last):
  File "/Users/jni/projects/napari-gui/napari_gui/components/_dimensions/view.py", line 87, in <lambda>
    self._slider_value_changed(value, axis))
  File "/Users/jni/projects/napari-gui/napari_gui/components/_dimensions/view.py", line 95, in _slider_value_changed
    self.dimensions._update()
  File "/Users/jni/projects/napari-gui/napari_gui/components/_dimensions/model.py", line 126, in _update
    self.viewer._update_layers()
  File "/Users/jni/projects/napari-gui/napari_gui/components/_viewer/model.py", line 207, in _update_layers
    self._update_status()
AttributeError: 'Viewer' object has no attribute '_update_status'
Abort trap: 6
```

Gonna try the notebook now but that error is probably worth investigating!

**Comment by jni** [comment]
**Edit:** when I run that file *and* try to move the dim slider.

**Comment by jni** [comment]
Behaviour that is probably unrelated to the PR but thought I'd bring it up anyway: if I click on the name of a layer that is unselected, it goes into text edit mode without selecting it. It would be nice if (a) it would select the layer, and/or (b) it wouldn't go into text editing mode until after the selection. This is analogous to e.g. file browsers like Finder, and it means that one can click and drag anywhere on the layer tile. Currently you can't click and drag if your click is on the name.

**Comment by jni** [comment]
Ok, in terms of the PR's functionality, it's super great. =D I agree with @freeman-lab's assessment: an alternative approach would be to have all these buttons "together" and always apparent, but greyed out for different layers. As @freeman-lab said, we can iterate on this based on user feedback.

And @sofroniewn you'll be shocked to hear I'm happy with the code! 😂 

Last comment: I hadn't really played around with the pan-zoom stuff since you went to holding space bar when in annotation mode. I was wondering why the "pan when click-drag on empty space" functionality was gone when I noticed the status bar helping me out and telling me to hold space bar! Massive usability win!

Side note about the annotation mode: probably we want to allow marker size control in that mode somehow... It feels a bit broken that marker sizes can be specified on the command line but not in the GUI. But this is very nitpicky and in no way urgent.

I'm going to merge this then call it a day — I'll make issues with the above comments tomorrow!

**Comment by jni** [comment]
Hmm, possibly a premature merge. I'd assumed that the `demo.py` failure was just bit rot, but I just saw my notebook console, and it's full of these:

```pytb
AttributeError: 'Viewer' object has no attribute '_update_status'
Traceback (most recent call last):
  File "/Users/jni/projects/napari-gui/napari_gui/components/_dimensions/view.py", line 87, in <lambda>
    self._slider_value_changed(value, axis))
  File "/Users/jni/projects/napari-gui/napari_gui/components/_dimensions/view.py", line 95, in _slider_value_changed
    self.dimensions._update()
  File "/Users/jni/projects/napari-gui/napari_gui/components/_dimensions/model.py", line 126, in _update
    self.viewer._update_layers()
  File "/Users/jni/projects/napari-gui/napari_gui/components/_viewer/model.py", line 207, in _update_layers
    self._update_status()
AttributeError: 'Viewer' object has no attribute '_update_status'
```

**Comment by sofroniewn** [comment]
@jni thanks for catching these, it seems like this was me not fully testing after every commit on (either the notebook or the demo.py) and missing the multidimensional use case. I'll get it fixed this morning and submit a followup PR.

Making issues with the other comments - layer selection on text edit / considerations around grayed out buttons / editing marker sizes in the gui - would be great too. Thanks!

**Pull request merged**